### DISCUSSION TITLE
vim-lsp 2022

### DISCUSSION THREAD

**Discussion by prabirshrestha** [prabirshrestha/vim-lsp#1255] [Ideas]
It is the time of the year to go back to the community and ask what you would like to see in vim-lsp going forward in 2022. I have done similar posts in the past [vim-lsp 2020](https://github.com/prabirshrestha/vim-lsp/issues/624) and [vim-lsp 2021 (poll)](https://github.com/prabirshrestha/vim-lsp/issues/1020). 

Please use one comment for each suggestion. If you want to expand on it use the replies. Use the thumbs-up and thumbs-down to express whether you agree with a specific answer/request (you can find that in the header of the comment on the top right, looks like a smiley face). Looking at the latest LSP spec we seem to have most of the features in vim-lsp except few big ones. There are also lots of possibilities for small user and ux improvements that I have in mind but as last year I want to see what you think is a top priority.

Last year we made a lot of perf improvements (not done yet), internal refactor, improved stability and added tons of new features. 

For the first time we opened up donations and had some contributions. If you are enjoying vim-lsp feel free to checkout https://opencollective.com/vim-lsp or individually donate to contributors of your choice.

Happy New Year!

**Comment by jmdevin** [comment]
Sometime this year, I think it would be wise to begin writing some features in vim9script. Vim9script can be [mixed](https://github.com/vim/vim/blob/475d9521baaeb060ebb915d678094ff28e67b5be/runtime/doc/vim9.txt#L1396) with legacy vimscript in the same file. While vim 9 itself has not been released yet, it shouldn't be too far off as Bram [indicates](https://groups.google.com/g/vim_dev/c/YXFyEJttWuM) the syntax and implementation are stable. The feature itself has been [added](https://github.com/vim/vim/commit/b79ee0c299d786627784f7304ba84b80e78ece26) already for testing. When vim 9 is eventually released and trickles down to the stable repos, this would result in a big performance gain for anyone who can't take advantage of lua.

**Reply by idbrii** [reply]
If stability is a goal of vim-lsp, then writing parts of it in vim9script before vim9 is even released seems premature. You might be able to mix vim9script with vimscript, but that means two implementations for every feature. We [already write vimscript bottlenecks in lua](https://github.com/prabirshrestha/vim-lsp/blob/296fb98d198cbbb5c5c937c09b84c8c7a9605a16/autoload/lsp/utils/diff.vim#L14-L40), so that's actually three implementations per feature.

**Reply by jmdevin** [reply]
This is a list of goals for 2022, not for January 3rd from what I understand, so I don't think it's premature when vim 9 will be out with plenty of time left in the year. Vim9script, unlike lua, will eventually ship with vim by default on popular distros. After that happens, there might not be a reason for bottlenecking code to be written in legacy vimscript at all, meaning only two implementations. Or one, if neovim decides to support vim9script.

**Reply by idbrii** [reply]
> there might not be a reason for bottlenecking code to be written in legacy vimscript at all

This assumes that vim-lsp should drop support for older versions of vim.

**Reply by jmdevin** [reply]
Eventually, I imagine that would happen, yes. Once vim 9 is the version you get on Ubuntu LTS, maybe. But it's certainly not required to add vim9script support in the first place. The unlikely event that neovim supports vim9script would certainly make things easier.

**Reply by Guergeiro** [reply]
I do agree with @jmdevin. vim9script would be a nice to be added, specially after it becomes the default on Debian/Ubuntu.

I know @prabirshrestha already said [he wouldn't support it](https://github.com/prabirshrestha/vim-lsp/issues/944#issuecomment-740774474), although personally I feel like the reason behind it is irrelevant since neovim now has its own lsp ecosystem.

**Reply by prabirshrestha** [reply]
I have not found the use of vim9script help improve perf yet, because most of the bottlenecks are better being implemented in vim core in c which will be even faster than implementing in vim9script.

For example:
* fuzzy searching: `matchfuzzy` was added to vim.
* lsp client in vim-lsp is quite slow. Vim now supports native LSP - https://github.com/vim/vim/issues/9390
* diff is slow but this has been fixed by using lua as mentioned above. I have added  [better lua](https://github.com/vim/vim/issues/6342) [support in vim](https://github.com/vim/vim/pull/6246).

If you have open source repositories that has perf issues feel free to share and I can profile and figure out the bottlenecks.

**Comment by 0xDagal** [comment]
As told [here](https://www.reddit.com/r/vim/comments/rwk5pp/vim_and_debuggers/hrfgtxk/), it could be nice to have a [DAP](https://microsoft.github.io/debug-adapter-protocol/) client, maybe in `vim-lsp`, maybe in another repo idk but it's sure that getting a debug adapter in pure VimScript would be really nice !

Note: why not, if it is possible, create [Machine Interface](https://ftp.gnu.org/old-gnu/Manuals/gdb/html_chapter/gdb_22.html) for [DAP](https://microsoft.github.io/debug-adapter-protocol/) so that we can use [termdebug](https://vimhelp.org/terminal.txt.html#terminal-debug) as client

**Reply by idbrii** [reply]
> getting a debug adapter in pure VimScript would be really nice

You mean something like [vimspector](https://github.com/puremourning/vimspector), but implemented only in vimscript? Note that vim-lsp isn't pure vimscript: [some parts are written in lua for speed](https://github.com/prabirshrestha/vim-lsp/blob/296fb98d198cbbb5c5c937c09b84c8c7a9605a16/autoload/lsp/utils/diff.vim#L15). But vim-lsp has vimscript fallbacks so it can run without requiring additional runtimes. Is that what you're after -- a more portable experience?

**Comment by prabirshrestha** [comment]
Lot of folks seems to want workspaceFolders support.

https://github.com/prabirshrestha/vim-lsp/issues/1069

**Reply by prabirshrestha** [reply]
Initial workspace folder support at https://github.com/prabirshrestha/vim-lsp/pull/1265

**Comment by zoumi** [comment]
I don't like coc.nvim. But for now coc.nvim return better complete items than vim-lsp.

**Reply by idbrii** [reply]
What's a specific example where coc returns better results? Can you point to a test file and a sample input line?

**Comment by mattn** [comment]
https://github.com/microsoft/language-server-protocol/issues/737

Explicitly handling for diagnostics.

**Reply by mattn** [reply]
Current implementation is based on notification from servers. So it might be heavily. But I think this spec change is possible to ommit some notifications if vim-lsp send diagnostics request with file save.

**Comment by egberts** [comment]
Also `let ` keyword has been eliminated from VimL9/vimscript9.
